proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2934,6 +2939,105 @@ lemma reply_remove_ccorres:
+  assumes "ts = BlockedOnReceive bo bicg ro"
+  shows "ro \<noteq> Some 0"
+  using assms
+  by (auto simp: valid_tcb_state'_def)
+
+lemma ctcb_relation_BlockedOnReceive_blockingObject:
+  assumes "ctcb_relation tcb ctcb"
+  assumes "BlockedOnReceive bo bicg ro = tcbState tcb"
+  shows "blockingObject_CL (thread_state_lift (tcbState_C ctcb)) = bo"
+  using assms
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma ctcb_relation_BlockedOnReceive_replyObject:
+  assumes "ctcb_relation tcb ctcb"
+  assumes "BlockedOnReceive bo bicg ro = tcbState tcb"
+  shows "replyObject_CL (thread_state_lift (tcbState_C ctcb)) = option_to_0 ro"
+  using assms
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma ctcb_relation_BlockedOnSend_blockingObject:
+  assumes "ctcb_relation tcb ctcb"
+  assumes "BlockedOnSend bo bib bicg bicgr biic = tcbState tcb"
+  shows "blockingObject_CL (thread_state_lift (tcbState_C ctcb)) = bo"
+  using assms
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma BlockedOnReceive_ep_at':
+  assumes "valid_objs' s"
+  assumes "st_tcb_at' ((=) (BlockedOnReceive bo bicg ro)) t s"
+  shows "ep_at' bo s"
+  using assms
+  apply -
+  apply (drule (1) tcb_in_valid_state')
+  by (fastforce simp: obj_at'_def valid_tcb_state'_def)
+
+lemma BlockedOnSend_ep_at':
+  assumes "valid_objs' s"
+  assumes "st_tcb_at' ((=) (BlockedOnSend bo bib bicg bicgr biic)) t s"
+  shows "ep_at' bo s"
+  using assms
+  apply -
+  apply (drule (1) tcb_in_valid_state')
+  by (fastforce simp: obj_at'_def valid_tcb_state'_def)
+
+lemma sym_ref_BlockedOnReceive_replyObject_linked:
+  assumes sy: "sym_refs (state_refs_of' s)"
+  assumes vo: "valid_objs' s"
+  assumes st: "st_tcb_at' ((=) (Structures_H.thread_state.BlockedOnReceive bo bicg (Some ro))) thread s"
+  shows "obj_at' (\<lambda>reply. replyTCB reply = Some thread) ro s"
+proof -
+  from vo and st have ra:
+    "reply_at' ro s"
+    by - (drule (1) tcb_in_valid_state', fastforce simp: obj_at'_def valid_tcb_state'_def)
+  from st have ko:
+    "\<exists>tcb. ko_at' tcb thread s \<and> tcbState tcb = BlockedOnReceive bo bicg (Some ro)"
+    by (clarsimp simp: st_tcb_at'_def obj_at'_def)
+  from sy and vo and st and ra and ko show ?thesis
+    apply -
+    apply (clarsimp simp: st_tcb_at'_def)
+    apply (frule (1) sym_ref_Receive_or_Reply_replyTCB')
+     apply simp
+    apply (clarsimp simp: obj_at'_def)
+    done
+qed
+
+lemma cancelIPC_threadSet_wp:
+  shows "threadSet (tcbFault_update (\<lambda>_. None)) t
+         \<lbrace>\<lambda>s. st_tcb_at' ((=) ts) t s \<and> invs' s
+              \<and> weak_sch_act_wf (ksSchedulerAction s) s \<and> sym_refs_asrt s\<rbrace>"
+proof -
+  have sy: "threadSet (tcbFault_update (\<lambda>_. None)) t \<lbrace>sym_refs_asrt\<rbrace>"
+    by - (unfold sym_refs_asrt_def, wpsimp wp: threadSet_state_refs_of', simp,
```

Comment:
if you're doing Isar-style proofs, it makes more sense to do:

```
unfolding sym_refs_asrt_def
by wpsimp
   (force simp: tcb_bound_refs'_def)+
```

Reply:
Done.